core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -1173,6 +1173,35 @@ class LogManager(logDirs: Seq[File],
     }
   }
 
+  def recoverAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Unit = {
+    val abandonedFutureLogs = findAbandonedFutureLogs(brokerId, newTopicsImage)
+    abandonedFutureLogs.foreach { log =>
+      val tp = log.topicPartition
+
+      log.renameDir(UnifiedLog.logDirName(tp), shouldReinitialize = true)
+      log.removeLogMetrics()
+      futureLogs.remove(tp)
+
+      currentLogs.put(tp, log)
+      log.newMetrics()
+
+      info(s"Successfully renamed abandoned future log for $tp")
+    }
+  }
+
+  private def findAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Iterable[UnifiedLog] = {
+    futureLogs.values.flatMap { log =>
+      val topicId = log.topicId.getOrElse {
+        throw new RuntimeException(s"The log dir $log does not have a topic ID, " +
+          "which is not allowed when running in KRaft mode.")
+      }
+      val partitionId = log.topicPartition.partition()
+      Option(newTopicsImage.getPartition(topicId, partitionId))
+        .filter(pr => directoryId(log.parentDir).contains(pr.directory(brokerId)))
+        .map(_ => log)
```

Comment:
1. Do we have the guarantee that the topic will be in the new topics image? Can't the topic be in later metadata delta if a compaction hasn't yet occurred?
2. Shouldn't we check if the main replica is on one of the other directories? If the main replica is on an offline dir (as in the scenario described in the issue) the broker will refuse to start once it is restarted with that directory online, as it will see the two main replicas. If the log directory for the main replica is online, we should be able to detect that here.

Reply:
Thanks for the feedback.

For (2), we've couple of options. We can either:

(a) ignore the future replica (say in dir2) if the main replica exists in an online log dir (say dir1) or,
(b) promote the future replica (in dir2)  and remove the main replica (in dir1).

(a) would result in ReplicaManager spawning a replicaAlterLogDir thread for the future replica and correcting the assignment to dir1, only for it to be changed back again to dir2 when the replicaAlterLogDir thread finishes its job. Refer https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2734 and https://github.com/apache/kafka/blob/acecd370cc3b25f12926e7a4664a2648f08c6c9a/core/src/main/scala/kafka/server/ReplicaManager.scala#L2745

Since in these scenarios, the future replica is almost caught up with the main replica, I'm leaning towards option (b) to avoid more reassignments. Please let me know if you feel otherwise.

Reply:
If we didn't know if the future log was caught up or not, then I'd prefer (a), but at this point I can't conceive of a different scenario – other than a failure during replica promotion – that would cause the future log to be in the directory assigned in the metadata. So I agree that the two logs likely will be either caught up or very close. So I agree it makes more sense to do (b) - promote the future log and delete the main one.

We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

Reply:
Addressed in 430546fa5a78118497ada6373607f6a25c78a8e8. Please take a look.

Reply:
I agree option (b) is better. But for this:
> We can still run into trouble if the directory with the main replica is offline. At some point that will cause a crash if the directory ever comes back online. But there's nothing we can do about that here. Maybe future work could improve how the broker handles loading conflicting logs.

I believe this PR: https://github.com/apache/kafka/pull/15335 should already fix this issue. @OmniaGM , could you help confirm it? If not, I think we should not promote the future log when main replica is offline to cause potential future issue.